File: hdf-converters/converters/hdf-to-xml/typescript/number.test.ts
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { describe, it, expect } from 'vitest';
import { convertHdfToXml } from './converter.js';

const __dirname = dirname(fileURLToPath(import.meta.url));

interface NumberCase {
  json: string;
  text: string;
  why: string;
}

const CASES = (
  JSON.parse(
    readFileSync(join(__dirname, '..', '..', '..', 'shared', 'xml-number-cases.json'), 'utf-8'),
  ) as { cases: NumberCase[] }
).cases;

/** The document is assembled as raw text: a JS number literal would already have
 * lost the distinction the table is pinning (-0 among them). */
function convert(rawNumber: string): string {
  return convertHdfToXml(
    `{"baselines":[{"name":"b","requirements":[{"id":"r","impact":0,` +
      `"tags":{"n":${rawNumber}},` +
      `"descriptions":[{"label":"default","data":"d"}],` +
      `"results":[{"status":"passed","codeDesc":"c","startTime":"2020-01-01T00:00:00Z"}]}]}]}`,
  );
}

// Rendering is implemented twice, so the expectations live in one shared file
// both languages read rather than in two hand-kept copies.
describe('hdf-to-xml number rendering', () => {
  it('has a populated shared table', () => {
    expect(CASES.length, 'an empty table would pass vacuously').toBeGreaterThan(0);
  });

  it.each(CASES.map((c) => [c.json, c] as const))('renders %s like the Go peer', (_json, c) => {
    expect(convert(c.json), c.why).toContain(`<n>${c.text}</n>`);
  });
});

File: hdf-converters/converters/hdf-to-xml/go/number_test.go
package hdftoxml

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

type numberCase struct {
	JSON string `json:"json"`
	Text string `json:"text"`
	Why  string `json:"why"`
}

func loadNumberCases(t *testing.T) []numberCase {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("..", "..", "..", "shared", "xml-number-cases.json"))
	require.NoError(t, err)

	var table struct {
		Cases []numberCase `json:"cases"`
	}
	require.NoError(t, json.Unmarshal(raw, &table))
	require.NotEmpty(t, table.Cases, "an empty table would pass vacuously")
	return table.Cases
}

// Rendering is implemented twice, so the expectations live in one shared file
// both languages read rather than in two hand-kept copies. This side passes
// today: it pins Go against future drift rather than proving the table correct,
// since the table was derived from this implementation and then checked against
// an independent shortest-round-trip renderer.
func TestConvertHDFToXMLNumberTextMatchesSharedTable(t *testing.T) {
	for _, c := range loadNumberCases(t) {
		t.Run(c.JSON, func(t *testing.T) {
			// Raw text, not a Go float: a literal would already have lost the
			// distinction the table is pinning (-0 among them).
			input := []byte(`{"baselines":[{"name":"b","requirements":[{"id":"r","impact":0,` +
				`"tags":{"n":` + c.JSON + `},` +
				`"descriptions":[{"label":"default","data":"d"}],` +
				`"results":[{"status":"passed","codeDesc":"c","startTime":"2020-01-01T00:00:00Z"}]}]}]}`)

			out, err := ConvertHDFToXML(input)
			require.NoError(t, err)
			assert.Contains(t, string(out), "<n>"+c.Text+"</n>", c.Why)
		})
	}
}
